python/ray/worker.py

```diff
@@ -459,13 +466,24 @@ def print_logs(self):
                     msg = subscriber.poll()
                 else:
                     msg = subscriber.get_message()
+                # GCS subscriber only returns None on unavailability.
+                # Redis subscriber returns None when there is no new message.
                 if msg is None:
                     num_consecutive_messages_received = 0
+                    last_polling_batch_size = 0
                     self.threads_stopped.wait(timeout=0.01)
                     continue
-                num_consecutive_messages_received += 1
-                if (num_consecutive_messages_received % 100 == 0
-                        and num_consecutive_messages_received > 0):
+
+                if self.gcs_pubsub_enabled:
```

Comment:
Instead of a heuristic, can we set a hard limit on the number of buffered log messages (say, 10MiB), and raise the warning if any are dropped from the buffer?

Btw, what currently happens if we can't keep up? Do we OOM the GCS?

Reply:
I think too much logs can have two issues (1) overwhelming the user, (2) overwhelming log processing. IIUC this heuristic was addressing (1). We can add protection for (2) separately.

Yes, currently too much logs / error info pushed to GCS would make GCS OOM, if any subscriber cannot keep up.